src/workerd/api/worker-rpc.h

```diff
@@ -47,7 +57,76 @@ class WorkerRpc : public Fetcher, public jsg::NamedIntercept {
   }
 };
 
-#define EW_WORKER_RPC_ISOLATE_TYPES  \
-  api::WorkerRpc
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<void>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext context) override;
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  kj::Own<kj::PromiseFulfiller<void>> callFulfiller;
+  IoContext& ctx;
+  kj::Maybe<kj::StringPtr> entrypointName;
+};
+
+class GetJsRpcTargetEvent final: public ExtendableEvent {
+public:
+  explicit GetJsRpcTargetEvent();
+
+  static jsg::Ref<GetJsRpcTargetEvent> constructor(kj::String type) = delete;
+
+  JSG_RESOURCE_TYPE(GetJsRpcTargetEvent) {
+    JSG_INHERIT(ExtendableEvent);
+  }
+};
+
+// `getJsRpcTarget` returns a capability that provides the client a way to call remote methods
+// over RPC. We drain the IncomingRequest after the capability is used to run the relevant JS.
+class GetJsRpcTargetCustomEventImpl final: public WorkerInterface::CustomEvent {
```

Comment:
nit: Can we move this impl to the source file?

Reply:
This one we can't because we reference it when allocating the custom event (it's hard to see because it's in the internal codebase).